ARCHITECTURE.md

```diff
@@ -0,0 +1,74 @@
+# Exonum Architecture
+
+This document describes the high-level architecture of Exonum. Its target
+audience is the developers of Exonum core. To learn how to *use* Exonum, see
+https://exonum.com/doc/.
+
+This document is provided on the best effort bases -- it is not guaranteed to be
+up-to-date and complete. The purpose of this document is to highlight the most
+important types, modules and functions in the code, to allow new contributors to
+dive into Exonum quickly. It is assumed that the reader is familiar with core
+concepts of Exonum, such as services and transactions.
+
+
+## Code organization
+
+The repository is a single Cargo workspace of several related packages:
+
+- `exonum`, most of the code lives here,
+
+- `sandbox`, where integration tests live,
+
+- `testkit`, a framework for testing Exonum services.
+
+
+This document currently describes only the `exonum` crate.
+
+
+## Storage
+
+Storage module can be found at `src/storage`. It provides an abstraction over an
+embedded key value store: the `Database` trait. The keys and values are just
+slices of bytes: `&[u8]`.
+
+To transform raw bytes into native Rust data structures, the `StorageKey` and
+`StorageValue` traits are used.
+
+On top of raw storage, collections like sets, lists and maps are provided: see
+various `*Index` structs. Of particular interest are `storage/proof_map_index`
+and `storage/proof_list_index` modules, which provide merkleized collections,
+capable of providing compact proofs for search queries.
+
+Indexes are used by the core of Exonum to store the blockchain state. Service
+implementors may use indexes to store service-specific data.
```

Comment:
I'd refer the reader to the built-in indexes, i.e., `blockchain/schema.rs`, right here rather than later. The notion that the core has built-in indexes is not trivial.